test/suite/wthooks.py

```diff
@@ -0,0 +1,216 @@
+                    self.hooks.append(hook)
+            except:
+                print('Cannot import hook: ' + name + ', check file ' + modname + '.py')
+                raise
+        for hook in self.hooks:
+            hook.setup_hooks()
+
+    def add_hook(self, clazz, method_name, hook_type, hook_func):
+        if not hasattr(clazz, method_name):
+            raise Exception('Cannot find method ' + method_name + ' on class ' + str(clazz))
+
+        # Given that the method name is XXXX, and class is Connection, here's what we're doing:
+        #    wiredtiger.Connection._XXXX_hooks = []
+        #    wiredtiger.Connection._XXXX_orig = wiredtiger.Connection.XXXX
+        #    wiredtiger.Connection.XXXX = lambda self, *args: hooked_function(self, 'XXXX', *args)
+        hooks_name = '_' + method_name + '_hooks'
+        orig_name = '_' + method_name + '_orig'
+        if not hasattr(clazz, hooks_name):
+            tty('Setting up hook on ' + str(clazz) + '.' + method_name)
+            setattr(clazz, hooks_name, [])
+            orig_func = getattr(clazz, method_name)
+            if orig_func == None:
+                raise Exception('method ' + method_name + ' hook setup: method does not exist')
+
+            # If we're using the wiredtiger module and not a class, we need a slightly different style of hooked_function.
+            if clazz == wiredtiger:
+                f = lambda *args: hooked_function(wiredtiger, orig_func, hooks_name, *args)
+            else:
+                f = lambda self, *args: hooked_function(self, orig_func, hooks_name, *args)
+            setattr(clazz, method_name, f)
+            setattr(clazz, orig_name, orig_func)
+
+        # Now add to the list of hook functions
+        # If it's a replace hook, we only allow one of them for a given method name
+        hooks_list = getattr(clazz, hooks_name)
+        if hook_type == HOOK_REPLACE:
+            for ht, hf in hooks_list:
+                if ht == HOOK_REPLACE:
+                    raise Exception('method ' + method_name + ' hook setup: trying to replace the same method with two hooks')
+        elif hook_type != HOOK_NOTIFY and hook_type != HOOK_ARGS:
+            raise Exception('method ' + method_name + ' hook setup: unknown hook_type: ' + str(hook_type))
+        hooks_list.append([hook_type, hook_func])
+        setattr(clazz, hooks_name, hooks_list)
+        tty('Setting up hooks list in ' + str(clazz) + '.' + hooks_name)
+
+    def get_function(self, clazz, method_name):
+        orig_name = '_' + method_name + '_orig'
+        if hasattr(clazz, orig_name):
+            orig_func = getattr(clazz, orig_name)
+        else:
+            orig_func = getattr(clazz, method_name)
+        return orig_func
+
+    def filter_tests(self, tests):
+        for hook in self.hooks:
+            tests = hook.filter_tests(tests)
+        return tests
+
+# Hooks must derive from this class
+class WiredTigerHook(object):
```

Comment:
Would converting this to an abstract base class be more beneficial? Enforcing that WiredTigerHook needs to be derived from another class

Reply:
Thanks - I did this.  My first time using "abc", so please check my work.